// homebridge-config-ui-x/dist/modules/status/status.gateway.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@nestjs/core");
const common_1 = require("@nestjs/common");
const websockets_1 = require("@nestjs/websockets");
const plugins_service_1 = require("../plugins/plugins.service");
const status_service_1 = require("./status.service");
const ws_guard_1 = require("../../core/auth/guards/ws.guard");
let StatusGateway = class StatusGateway {
    constructor(statusService, pluginsService, moduleRef) {
        this.statusService = statusService;
        this.pluginsService = pluginsService;
        this.moduleRef = moduleRef;
    }
    serverStatus(client, payload) {
        return __awaiter(this, void 0, void 0, function* () {
            this.statusService.watchStats(client);
        });
    }
    homebridgeVersionCheck(client, payload) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                return yield this.pluginsService.getHomebridgePackage();
            }
            catch (e) {
                return new websockets_1.WsException(e.message);
            }
        });
    }
    getOutOfDatePlugins(client, payload) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                return yield this.pluginsService.getOutOfDatePlugins();
            }
            catch (e) {
                return new websockets_1.WsException(e.message);
            }
        });
    }
};
__decorate([
    websockets_1.SubscribeMessage('monitor-server-status'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], StatusGateway.prototype, "serverStatus", null);
__decorate([
    websockets_1.SubscribeMessage('homebridge-version-check'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], StatusGateway.prototype, "homebridgeVersionCheck", null);
__decorate([
    websockets_1.SubscribeMessage('get-out-of-date-plugins'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], StatusGateway.prototype, "getOutOfDatePlugins", null);
StatusGateway = __decorate([
    common_1.UseGuards(ws_guard_1.WsGuard),
    websockets_1.WebSocketGateway({ namespace: 'status' }),
    __metadata("design:paramtypes", [status_service_1.StatusService,
        plugins_service_1.PluginsService,
        core_1.ModuleRef])
], StatusGateway);
exports.StatusGateway = StatusGateway;
